File: bots/beacon/bestrn/builder/tasks/rejected.py
"""
Translation of `bots/intgrah/v54.7.9/builder/tasks/rejected.py`.

Failure model is Option-shape, not Result: `None` = task fired,
`Some(TaskRejected)` = task rejected. Dodges Python exceptions
(which would be non-deterministic if any unrelated exception
propagated mid-turn) and matches the pyrust DSL `try_!` macro for
early propagation.
"""
from __future__ import annotations

class TaskRejected:
    reason: str

    def __init__(self, reason):
        """Build a rejection from a static reason string."""
        self.reason = str(reason)

    @staticmethod
    def from_string(reason):
        """Build a rejection from an owned reason string (for templated reasons)."""
        __self = TaskRejected.__new__(TaskRejected)
        __self.reason = reason
        return __self

    def fmt(self, f):
        return f.write_str(self.reason)
type TaskResult = TaskRejected | None

File: bots/rs/v56/src/builder/tasks/rejected.rs
//! Translation of `bots/intgrah/v54.7.9/builder/tasks/rejected.py`.
//!
//! Failure model is Option-shape, not Result: `None` = task fired,
//! `Some(TaskRejected)` = task rejected. Dodges Python exceptions
//! (which would be non-deterministic if any unrelated exception
//! propagated mid-turn) and matches the pyrust DSL `try_!` macro for
//! early propagation.

#[derive(Clone, Debug)]
pub struct TaskRejected {
    pub reason: String,
}

impl TaskRejected {
    /// Build a rejection from a static reason string.
    #[must_use]
    pub fn new(reason: &str) -> Self {
        Self {
            reason: pyrust::to_string!(reason),
        }
    }

    /// Build a rejection from an owned reason string (for templated reasons).
    #[must_use]
    pub const fn from_string(reason: String) -> Self {
        Self { reason }
    }
}

impl core::fmt::Display for TaskRejected {
    #[pyrust::inline]
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        f.write_str(&self.reason)
    }
}

/// Result type used by every task leaf: `None` = task fired,
/// `Some(TaskRejected)` = move on to the next sibling.
pub type TaskResult = Option<TaskRejected>;
